### ISSUE TITLE
Bug: converting stix messages with DocumentUtilities.toDocument is very slow( > 1 second)

### ISSUE THREAD

**Issue by manonthegithub** [STIXProject/java-stix#12]
The method in DocumentUtilities has following implementation:
```java
public static Document toDocument(JAXBElement<?> jaxbElement,
			boolean prettyPrint) {

		Document document = null;
		
		try {
			DocumentBuilderFactory documentBuilderFactory = DocumentBuilderFactory
					.newInstance();
			documentBuilderFactory.setNamespaceAware(true);
			documentBuilderFactory.setIgnoringElementContentWhitespace(true);
			documentBuilderFactory.isIgnoringComments();
			documentBuilderFactory.setCoalescing(true);

			document = documentBuilderFactory.newDocumentBuilder()
					.newDocument();
//!!! this is costly operation, should be performed once per application!!!
			JAXBContext jaxbContext = JAXBContext.newInstance(jaxbElement
					.getDeclaredType().getPackage().getName());

			Marshaller marshaller = jaxbContext.createMarshaller();

			marshaller.setProperty(Marshaller.JAXB_FORMATTED_OUTPUT,
					prettyPrint);

			marshaller.setProperty(Marshaller.JAXB_ENCODING, "UTF-8");

			try {
				marshaller.marshal(jaxbElement, document);
			} catch (JAXBException e) {
				// otherwise handle non-XMLRootElements
				QName qualifiedName = new QName(
						STIXSchema.getNamespaceURI(jaxbElement), jaxbElement
								.getClass().getSimpleName());

				@SuppressWarnings({ "rawtypes", "unchecked" })
				JAXBElement root = new JAXBElement(qualifiedName,
						jaxbElement.getClass(), jaxbElement);

				marshaller.marshal(root, document);
			}

			removeUnusedNamespaces(document);

		} catch (ParserConfigurationException e) {
			throw new RuntimeException(e);
		} catch (JAXBException e) {
			throw new RuntimeException(e);
		}
		
		return document;
	}
```

The problem is that new JAXBContext is initialized with JAXBContext.newInstance every time you want to convert your stix package to document, but it is costly and takes more than a second. JAXBContext should be instantiated once and cached.

**Comment by ShyanFung** [comment]
I stubbled upon your reported issue and found it to be very helpful.  

I noticed java8 was taking a very long time and lots of resources with the "JAXBContext" code you pointed out.  In my code, this was being called many times because I was adding/wrapping the STIX content block/object within my TAXII response message(s).

I added the following "helper" variable and method (plus minor update to toDocument to call the added function) to that DocumentUtilities class to help address the jaxb caching with consideration on threading.

```
	private static final Map<String, ThreadLocal<JAXBContext>> STIX_JAXB_CONTEXTS = 
			new ConcurrentHashMap<String, ThreadLocal<JAXBContext>>();

	public static Marshaller getStixJaxbMarshaller(String packageName) {
		ThreadLocal<JAXBContext> tljc = STIX_JAXB_CONTEXTS.get(packageName);
		if (tljc == null) {
			synchronized(DocumentUtilities.class) {
				try {
					JAXBContext jaxbContext = JAXBContext.newInstance(packageName);
					tljc = new ThreadLocal<JAXBContext>();
					tljc.set(jaxbContext);
					STIX_JAXB_CONTEXTS.put(packageName, tljc);
				} catch (Exception e) {
					LOGGER.warn("Cannot create JAXB context for : " + packageName);
					return null;
				}
			}
		}

		try {
			return tljc.get().createMarshaller();
		} catch (JAXBException e) {
			LOGGER.warn("Cannot create JAXB marshaller : " + e.getMessage());
			return null;
		}
	}
```

After adding my change, a response of 500 embedded STIX observables went from taking over a minute to several seconds.

**Comment by manonthegithub** [comment]
@ShyanFung You can use the only instance and remove part with `ThreadLocals` because JaxbContext is [thread safe ](https://stackoverflow.com/questions/7400422/jaxb-creating-context-and-marshallers-cost). This will improve you performance even more and make code more readable. In my app I made the only final instance of it and wrote my own `toDocument` conversions.

**Comment by manonthegithub** [comment]
Unfortunately, as I can see this lib is no longer maintained.

**Comment by nemonik** [comment]
Submit a pull request with your improvements with code to exercise and will merge and release...

**Comment by manonthegithub** [comment]
@nemonik ok, I will submit)

**Comment by manonthegithub** [comment]
@nemonik It is not clear how to do this optimization in current codebase. As for my project I just completely eliminated `DocumentUtilities.toDocument` and `DocumentUtilities.toXmlString` (it uses toDocument) and instantiated cached JaxbContext and then used marshallers and unmarshallers instantiated from it. 
Currently I do not have much free time, so will try to make some improvements and submit PR in a week or two.

**Comment by manonthegithub** [comment]
@ShyanFung you may try improved version in the PR and compare performance in your project).
@nemonik waiting for your comments)

**Comment by nemonik** [comment]
Will visit ASAP.

**Comment by ShyanFung** [comment]
@manonthegithub I checked out your change visually (in your pull request) and running it locally through API/app.  It looks good and performs a lot faster with a less memory and cpu overhead (the improvement is very obvious w/o having to benchmark it). 

I had taken your suggestion about JAXBContext already being thread-safe, and did my own workaround.

As side note, I moved away from the model of embedding many STIX packages per single TAXII content block.  This model was causing OOM errors due to marshaling many STIX DOM objects (in-memory).  My API now just wraps many STIX indicators per STIX package per single TAXII content block.  This message style reduced the number of concurrent calls to STIX's DocumentUtilities.toDocument calls.  Additionally, I added defaulting support for multi-part poll exchange to break up the message's content block.

Your pull request change will definitely help as more concurrent requests are made to the runtime.  Thanks again for the find and suggestions.
